### PR TITLE
feat(KL-119): add s3 image

### PR DESCRIPTION
**Pull request by ohhamma**
## 📌 연관된 이슈
- **[KL-119/상품사진 s3 등록](https://ohhamma.atlassian.net/browse/KL-119)**

## 📝 작업 내용
- s3 연결 (aws sdk v2 사용)
  - 사진 업로드 시 presigned url 반환
  - 업로드 완료 시 상품 / 유저 사진 필드 업데이트 (기존 사진은 deprecated 처리)
- cloudfront 연결
  - 사진 조회 시 cloudfront 도메인으로 조회하도록

## 🌳 작업 브랜치명
- `KL-119/상품사진-s-3-등록`

## Summary by CodeRabbit

- **New Features**
	- Introduced functionality for managing image uploads, including presigned URL generation for user and product images.
	- Added a new endpoint for updating user information in the UserController.
	- Enhanced product management with dynamic handling of associated images.

- **Bug Fixes**
	- Updated user retrieval logic to provide more accurate user data handling and error management.

- **Refactor**
	- Simplified the creation and updating of user and product entities by restructuring methods.
	- Improved the readability and maintainability of the codebase by isolating specific functionalities into dedicated methods.

- **Documentation**
	- Updated API metadata for improved clarity in documentation.

- **Chores**
	- Improved error handling capabilities by adding specific exceptions related to image processing and user operations.

### CODE REVIEW

**Comment by ohhamma** [comment]
@coderabbitai review

**Comment by ohhamma** [comment]
@coderabbitai review

File: src/main/resources/application-storage.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  aws:
+    credentials:
+      access-key: ${S3_ACCESS_KEY}
+      secret-key: ${S3_SECRET_KEY}
+    region:
+      static: ${S3_REGION}
+    s3:
+      bucket: ${S3_BUCKET_NAME}
+    cloudfront:
+      domain: ${S3_CLOUDFRONT_DOMAIN}
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/resources/application-storage.yaml]: got it

File: src/main/java/taco/klkl/domain/user/domain/Gender.java [hunk 2]
```diff
@@ -4,22 +4,21 @@
...
+	private final String value;
 
-	public static Gender getGenderByDescription(String description) {
+	public static Gender from(final String value) {
 		return Arrays.stream(Gender.values())
-			.filter(g -> g.getDescription().equals(description))
+			.filter(gender -> gender.getValue().equals(value))
 			.findFirst()
-			.orElse(NONE);
+			.orElseThrow(GenderNotFoundException::new);
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/domain/user/domain/Gender.java:22]: i'll add it later

File: src/main/java/taco/klkl/domain/image/domain/FileExtension.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+	private final String value;
+
+	public static FileExtension from(final String fileExtension) throws FileExtensionNotFoundException {
+		return Arrays.stream(FileExtension.values())
+			.filter(extension -> extension.toString().equals(fileExtension))
+			.findFirst()
+			.orElseThrow(FileExtensionNotFoundException::new);
+	}
+}
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/domain/image/domain/FileExtension.java]: got it. thanks.

File: src/main/java/taco/klkl/domain/product/dto/response/ProductSimpleResponse.java [hunk 4]
```diff
@@ -1,34 +1,30 @@
...
 ) {
-
 	public static ProductSimpleResponse from(final Product product) {
+		List<ProductImageResponse> images = product.getImages().stream()
+			.map(ProductImageResponse::from)
+			.toList();
+
 		return new ProductSimpleResponse(
 			product.getId(),
+			images,
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/domain/product/dto/response/ProductSimpleResponse.java:27]: i'll add it later

File: src/main/java/taco/klkl/global/config/s3/S3Config.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+public class S3Config {
+
+	@Value("${cloud.aws.credentials.access-key}")
+	private String accessKey;
+
+	@Value("${cloud.aws.credentials.secret-key}")
+	private String secretKey;
+
+	@Value("${cloud.aws.region.static}")
+	private String region;
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/global/config/s3/S3Config.java:25]: it's okay.

File: src/main/java/taco/klkl/global/util/UserUtil.java [hunk 6]
```diff
@@ -29,7 +30,7 @@ public User findTestUser() {
 	 * @return
 	 */
 	public User findCurrentUser() {
-		return userRepository.findFirstByName(UserConstants.TEST_USER_NAME);
+		return findTestUser();
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/global/util/UserUtil.java:33]: i'll add it later

File: src/main/java/taco/klkl/domain/product/service/ProductServiceImpl.java [hunk 7]
```diff
@@ -287,6 +284,13 @@ private void updateProductEntity(final Product product, final ProductCreateUpdat
 		);
 	}
 
+	private void updateProductEntityTags(final Product product, final Set<Long> tagIds) {
+		if (tagIds != null) {
+			final Set<Tag> updatedTags = createTagsByTagIds(tagIds);
+			product.updateTags(updatedTags);
+		}
+	}
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/domain/product/service/ProductServiceImpl.java:292]: i checked that condition in `product.updateTags(updatedTags)`

File: src/main/java/taco/klkl/domain/product/domain/Product.java [hunk 8]
```diff
@@ -252,4 +238,37 @@ public int decreaseLikeCount() throws LikeCountBelowMinimumException {
...
 		return this.likeCount;
 	}
+
+	public void updateImages(final List<String> imageUrls) {
+		this.images.clear();
+		IntStream.range(0, imageUrls.size())
+			.forEach(i -> {
+				ProductImage newImage = ProductImage.of(this, imageUrls.get(i), i);
+				this.images.add(newImage);
+			});
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/domain/product/domain/Product.java:248]: the method that calls this method has `@Transactional` annotation. i assume it's okay.

[the remark this answers is not in the document]
Comment (ohhamma) [reply src/main/java/taco/klkl/domain/product/domain/Product.java:248]: the method that calls this method has `@Transactional` annotation. i assume it's okay.

File: src/test/java/taco/klkl/domain/notification/service/NotificationServiceTest.java [hunk 9]
```diff
@@ -76,7 +76,7 @@ public class NotificationServiceTest {
 
 	@BeforeEach
 	public void setUp() {
-		commentUser = User.of("profile",
+		commentUser = User.of(
```
[the remark this answers is not in the document]
Comment (ohhamma) [reply src/test/java/taco/klkl/domain/notification/service/NotificationServiceTest.java:79]: i'll add it later

**Review by idealflower-k — APPROVED** [review]
팀장님 이 어려운 걸 혼자 다 하시다니 대단하십니다...
3S와 기타 수정사항 확인했습니다!
고생하셨습니다! :)

**Review by min3m — APPROVED** [review]
커밋량이랑 파일 갯수가 정말 어머어마 하네요…! 저랑 상화님이 빠져서 혼자 다 하셨어야했을 텐데 정말 고생많으셨습니다!!! S3관련 커밋 확인했습니다! 고생하셨습니다~~

**Pull request merged**